src/core/mavsdk_impl.cpp

```diff
@@ -110,8 +110,71 @@ std::vector<std::shared_ptr<System>> MavsdkImpl::systems() const
     return systems_result;
 }
 
-void MavsdkImpl::receive_message(mavlink_message_t& message)
+void MavsdkImpl::forward_message(mavlink_message_t& message, Connection* connection)
+{
+    /**
+     * @brief forward_message Function implementing Mavlink routing rules.
+     * See https://mavlink.io/en/guide/routing.html
+     *
+     * This function was adapted from PX4 Firmware v1.11.3
+     * https://github.com/PX4/PX4-Autopilot/blob/v1.11.3/src/modules/mavlink/mavlink_main.cpp
+     */
+    const mavlink_msg_entry_t* meta = mavlink_get_msg_entry(message.msgid);
+
+    bool forward_heartbeats_enabled = false;
+    int target_system_id = 0;
+    int target_component_id = 0;
+
+    // might be nullptr if message is unknown
+    if (meta) {
+        // Extract target system and target component if set
+        if (meta->flags & MAV_MSG_ENTRY_FLAG_HAVE_TARGET_SYSTEM) {
+            target_system_id = (_MAV_PAYLOAD(&message))[meta->target_system_ofs];
+        }
+
+        if (meta->flags & MAV_MSG_ENTRY_FLAG_HAVE_TARGET_COMPONENT) {
+            target_component_id = (_MAV_PAYLOAD(&message))[meta->target_component_ofs];
+        }
+    }
+
+    // If it's a message only for us, we keep it, otherwise, we forward it.
+    const bool targeted_only_at_us =
+        (target_system_id == get_own_system_id() && target_component_id == get_own_component_id());
+
+    // We don't forward heartbeats unless it's specifically enabled.
+    const bool heartbeat_check_ok =
+        (message.msgid != MAVLINK_MSG_ID_HEARTBEAT || forward_heartbeats_enabled);
+
+    if (!targeted_only_at_us && heartbeat_check_ok) {
+        std::lock_guard<std::mutex> lock(_connections_mutex);
+
+        uint8_t successful_emissions = 0;
```

Comment:
Sorry, I might have confused you here. I didn't mean that we need to specifically try to use short types, just the normal `unsigned` is fine. I just didn't know that `uint` is a thing.

```suggestion
        unsigned successful_emissions = 0;
```

Reply:
No problem, but it was OK for me as they might not be more than 255 connections ?

Reply:
Oh, because the map is limited to 255 as well, right?

Reply:
Hmm 🤔 I think I just wanted to keep it as small as possible, but it was arbitrary.
For the map, you meant the vector of Connection ? It does not seem to be limited to 255.

Reply:
I will just change it to `unsigned` :+1: